File: statistics.ml


let mean lst =
  let sum = List.fold_left (+) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let len = List.length sorted in
  if len mod 2 = 1 then
    float_of_int (List.nth sorted (len / 2))
  else
    let mid1 = List.nth sorted (len / 2 - 1)
    and mid2 = List.nth sorted (len / 2) in
    (float_of_int (mid1 + mid2)) /. 2.0

let mode lst =
  let count_map =
    List.fold_left (fun acc x ->
      if List.mem_assoc x acc then
        (x, List.assoc x acc + 1) :: List.remove_assoc x acc
      else
        (x, 1) :: acc
    ) [] lst in
  let max_count = List.fold_left (fun acc (_, c) -> max acc c) 0 count_map in
  List.fold_right (fun (k, v) acc -> if v = max_count then k :: acc else acc) count_map []

let () =
  let data = [1; 2; 2; 3; 4; 4; 5] in
  Printf.printf "Mean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);
  Printf.printf "Mode: ";
  List.iter (fun x -> Printf.printf "%d " x) (mode data);
  print_newline ()

File: statistics.c
// statistics.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

double calculate_mean(int arr[], int size) {
    int sum = 0;
    for (int i = 0; i < size; i++) sum += arr[i];
    return (double)sum / size;
}

double calculate_median(int arr[], int size) {
    qsort(arr, size, sizeof(int), compare);
    if (size % 2 == 0)
        return (arr[size / 2 - 1] + arr[size / 2]) / 2.0;
    else
        return arr[size / 2];
}

void calculate_mode(int arr[], int size) {
    int max_count = 0;
    int mode[size], mode_count = 0;
    
    for (int i = 0; i < size; i++) {
        int count = 0;
        for (int j = 0; j < size; j++)
            if (arr[j] == arr[i]) count++;
        
        if (count > max_count) {
            max_count = count;
            mode[0] = arr[i];
            mode_count = 1;
        } else if (count == max_count && arr[i] != mode[mode_count - 1]) {
            mode[mode_count++] = arr[i];
        }
    }

    printf("Mode: ");
    for (int i = 0; i < mode_count; i++) {
        printf("%d ", mode[i]);
    }
    printf("\n");
}

int main() {
    int data[] = {1, 2, 2, 3, 4, 4, 5};
    int size = sizeof(data) / sizeof(data[0]);

    printf("Mean: %.2f\n", calculate_mean(data, size));
    printf("Median: %.2f\n", calculate_median(data, size));
    calculate_mode(data, size);

    return 0;
}
